### PR TITLE
Fortify: Add map area markers for ACEX Fortify Objects

### PR DESCRIPTION
**Pull request by Seb105**
**When merged this pull request will:**
- Add map markers to ACEX Fortify placed static objects
- Configurable via CBA: Always, Never, Only show markers for friendlies
- Removed with placed object deletion

### CODE REVIEW

File: addons/fortify/XEH_postInit.sqf [hunk 1]
```diff
@@ -7,6 +7,7 @@ if (isServer) then {
         TRACE_3("objectPlaced",_unit,_side,_object);
         private _jipID = [QGVAR(addActionToObject), [_side, _object]] call CBA_fnc_globalEventJIP;
         [_jipID, _object] call CBA_fnc_removeGlobalEventJIP; // idealy this function should be called on the server
+        if (GVAR(markObjectsOnMap) isNotEqualTo 0 && {_object isKindOf "Static"}) then {[_object] call FUNC(createObjectMarker)};
```
Comment (jonpas) [inline addons/fortify/XEH_postInit.sqf:10]:
```suggestion
        if (GVAR(markObjectsOnMap) isNotEqualTo 0 && {_object isKindOf "Static"}) then {
            [_object] call FUNC(createObjectMarker)
        };
```

File: addons/fortify/XEH_postInit.sqf [hunk 2]
```diff
@@ -63,3 +64,18 @@ GVAR(objectRotationZ) = 0;
...
     };
 }] call CBA_fnc_addEventHandler;
+
+[QGVAR(setMarkerVisible), {
+    params ["_object"];
+    [
+        {!isNull player},
+        {
+            params ["_object"];
+            if (GVAR(markObjectsOnMap) isEqualTo 1 || {[_object getVariable QGVAR(objectSide), side group player] call BIS_fnc_sideIsEnemy}) then {
```
Comment (jonpas) [inline addons/fortify/XEH_postInit.sqf:74]: Might prefer [`side1 getFriend side2 < 0.6`](https://community.bistudio.com/wiki/getFriend), overhead of BIS functions is usually not great.

File: addons/fortify/functions/fnc_createObjectMarker.sqf [hunk 3]
```diff
@@ -0,0 +1,48 @@
+#include "script_component.hpp"
+
+/*
+ * Author: Seb
+ * Creates a map marker for a created static object but only for sides friendly to the creator side
```
Comment (jonpas) [inline addons/fortify/functions/fnc_createObjectMarker.sqf:5]:
```suggestion
 * Creates a map marker for a created static object but only for sides friendly to the creator side.
```

File: addons/fortify/functions/fnc_createObjectMarker.sqf [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * Arguments:
+ * 0: Created fortify object <OBJECT>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * _object call acex_fortify_fnc_createObjectMarker
+ *
+ * Public: Yes
```
Comment (jonpas) [inline addons/fortify/functions/fnc_createObjectMarker.sqf:16]: This function doesn't do enough input checking to be considered as public API in my opinion.

File: addons/fortify/functions/fnc_createObjectMarker.sqf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+private _p2 = _bbr select 1;
+private _maxWidth = abs ((_p2 select 0) - (_p1 select 0));
+private _maxLength = abs ((_p2 select 1) - (_p1 select 1));
+private _direction = getDir _object;
+
+// Marker name unique to this object
+private _markerNameStr = format ["acex_fortify_marker_%1", _object];
+
+// Create marker, set alpha using global event
+private _marker = createMarker [_markerNameStr, _object];
```
Comment (jonpas) [inline addons/fortify/functions/fnc_createObjectMarker.sqf:33]:
```suggestion
private _marker = createMarkerLocal [_markerNameStr, _object];
```
Same optimization as you did with the other commands.

File: addons/fortify/initSettings.sqf [hunk 6]
```diff
@@ -9,3 +9,17 @@
         2
     ]
 ] call CBA_settings_fnc_init;
+[
```
Comment (jonpas) [inline addons/fortify/initSettings.sqf:12]:
```suggestion

[
```

File: addons/fortify/initSettings.sqf [hunk 7]
```diff
@@ -9,3 +9,17 @@
...
+    LLSTRING(settingsCategory),
+    [
+        [0, 1, 2],
+        [LLSTRING(markObjectsOnMapNone), LLSTRING(markObjectsOnMapFriendly), LLSTRING(markObjectsOnMapEveryone)],
+        1
+    ],
+    true,
+    {},
+    true
+] call CBA_settings_fnc_init;
```
Comment (jonpas) [inline addons/fortify/initSettings.sqf:25]:
```suggestion
] call CBA_fnc_addSetting;
```

File: addons/fortify/stringtable.xml [hunk 8]
```diff
@@ -155,6 +155,21 @@
             <Russian>Показывать всегда</Russian>
             <Turkish>Her Zaman Göster</Turkish>
         </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMap">
+            <English>Create map markers</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapDesc">
+            <English>Controls when budget update hints are shown</English>
```
Comment (jonpas) [inline addons/fortify/stringtable.xml:162]:
```suggestion
            <English>Controls when budget update hints are shown.</English>
```

File: addons/fortify/functions/fnc_createObjectMarker.sqf [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+// Get Object size and direction
+private _bbr = 0 boundingBoxReal _object;
+private _p1 = _bbr select 0;
+private _p2 = _bbr select 1;
+private _maxWidth = abs ((_p2 select 0) - (_p1 select 0));
+private _maxLength = abs ((_p2 select 1) - (_p1 select 1));
+private _direction = getDir _object;
+
+// Marker name unique to this object
+private _markerNameStr = format ["acex_fortify_marker_%1", _object];
```
Comment (jonpas) [inline addons/fortify/functions/fnc_createObjectMarker.sqf:30]:
```suggestion
private _markerNameStr = format [QGVAR(marker_%1), _object];
```

File: addons/fortify/XEH_postInit.sqf [hunk 10]
```diff
@@ -63,3 +64,18 @@ GVAR(objectRotationZ) = 0;
...
+                private _marker = _object getVariable QGVAR(mapMarker);
+                _marker setMarkerAlpha 0;
+            };
+        }, 
+        _object
+    ] call CBA_fnc_waitUntilAndExecute;
```
Comment (jonpas) [inline addons/fortify/XEH_postInit.sqf:80]: There must be a better way to do this.

**Review by jonpas — CHANGES_REQUESTED** [review]

File: addons/fortify/functions/fnc_createObjectMarker.sqf (hunk 4 above)
Comment (Seb105) [reply addons/fortify/functions/fnc_createObjectMarker.sqf:16]: Apologies, I thought that Public meant a Global result.

Comment (jonpas) [reply addons/fortify/functions/fnc_createObjectMarker.sqf:16]: No worries, that's why there are reviews. :)

File: addons/fortify/initSettings.sqf (hunk 7 above)
Comment (Seb105) [reply addons/fortify/initSettings.sqf:25]: Same file is also using CBA_settings_fnc_init, isn't mixing & matching worse?

File: addons/fortify/XEH_postInit.sqf (hunk 10 above)
Comment (Seb105) [reply addons/fortify/XEH_postInit.sqf:80]: Gotta wait somehow for the player object to exist for JIP side checking.

File: addons/fortify/initSettings.sqf (hunk 7 above)
Comment (jonpas) [reply addons/fortify/initSettings.sqf:25]: Can also change the others, here or in separate PR. addSetting is public API, init is old.

File: addons/fortify/XEH_postInit.sqf [hunk 11]
```diff
@@ -63,3 +66,21 @@ GVAR(objectRotationZ) = 0;
...
 }] call CBA_fnc_addEventHandler;
+
+[QGVAR(setMarkerVisible), {
+    params ["_object"];
+    [
+        {!isNull player},
+        {
+            params ["_object"];
+            private _objectSide = _object getVariable QGVAR(objectSide);
+            private _playerSide = side group player;
```
Comment (mharis001) [inline addons/fortify/XEH_postInit.sqf:77]: What happens if the player changes side mid-mission or remote controls a unit?

File: addons/fortify/stringtable.xml [hunk 12]
```diff
@@ -155,6 +155,21 @@
...
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapNone">
+            <English>Never</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapFriendly">
+            <English>For friendlies only</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapEveryone">
+            <English>For everyone</English>
+        </Key>
```
Comment (mharis001) [inline addons/fortify/stringtable.xml:172]: Title case for display names and "an ACEX static object" -> "a fortify object" or "a fortification".

File: addons/fortify/XEH_postInit.sqf (hunk 11 above)
Comment (Seb105) [reply addons/fortify/XEH_postInit.sqf:77]: TeamSwitch EH + CBA_fnc_addPlayerEventHandler covers player side change or RC, I can add that.

I do not believe there is a way to detect side relation changes (i.e setFriend) without a constant check.

Comment (mharis001) [reply addons/fortify/XEH_postInit.sqf:77]: Which `CBA_fnc_addPlayerEventHandler`? I do not see one added by this PR. Also, there is a CBA player event handler for `"group"`.

Comment (Seb105) [reply addons/fortify/XEH_postInit.sqf:77]: Typo, I meant to say "could cover". I will be implementing those 2 player EHs to recalculate the desired alphas when the current unit/group changes.

I've done some digging and I still don't see a way to handle side relation changes.

**Pull request closed without merging**